### PR TITLE
devhost: make container cleanup optional, make nginx more robust

### PR DESCRIPTION
**Pull request by ctheune**
@flyingcircusio/release-managers

## Release process

Impact:

none

Changelog:

* Improve our container `devhost` role: automatic cleanup is optional now and nginx configs are more robust regarding container DNS lookups.

## Security implications

- [x] [Security requirements](https://wiki.flyingcircus.io/System_Development_Guideline#Security_requirement_principles_and_testing) defined? (WHERE)

no specific requirements afaict

- [x] Security requirements tested? (EVIDENCE)

tested manually on our internal devhost

### CODE REVIEW

File: nixos/roles/devhost/default.nix [hunk 1]
```diff
@@ -8,13 +8,12 @@ let
...
-              (filename: hasSuffix ".json" filename)
-              (attrNames (readDirMaybe "/etc/devhost/")))))
+  allContainers = with lib;
+     (map
+      (filename: fromJSON (readFile "/etc/devhost/${filename}"))
+      (filter
+        (filename: hasSuffix ".json" filename)
+        (attrNames (readDirMaybe "/etc/devhost/"))))
     # BBB can be removed after a grace period
     ++ (filter (container: container.enabled or true)
```
Comment (dpausp) [inline nixos/roles/devhost/default.nix]: Does it make sense to filter for enabled containers here? The result is called `allContainers`

File: nixos/roles/devhost/default.nix [hunk 2]
```diff
@@ -191,6 +212,16 @@ in
       };
       networking.nat.internalInterfaces = ["ve-+"];
 
+      networking.extraHosts = ''
+        # static entries for devhost containers to avoid nginx issues
+        # if containers are not running.
+        #
+        # This could be a problem with duplicate entries, but I think
+        # we'll fine.
```
Comment (dpausp) [inline nixos/roles/devhost/default.nix]: "we'll be fine"

File: nixos/roles/devhost/default.nix [hunk 3]
```diff
@@ -181,6 +190,18 @@ in
         fi
       '';
 
+      # BBB can be removed after a grace period
+      flyingcircus.activationScripts.devhost-migrate-ip-config = ''
+        set -x
+        cd /etc/devhost
```
Comment (dpausp) [inline nixos/roles/devhost/default.nix:195]: Is it ok to ignore the old `/etc/devserver` here?

**Review by dpausp — CHANGES_REQUESTED** [review]

File: nixos/roles/devhost/default.nix (hunk 3 above)
Comment (ctheune) [reply nixos/roles/devhost/default.nix:195]: Yeah, this has been rolled out already and I'm assuming the other script runs first in any case. Also, even if it doesn't it's working in a converging manner, so ...

File: nixos/roles/devhost/default.nix (hunk 1 above)
Comment (ctheune) [reply nixos/roles/devhost/default.nix]: Stupid copy and paste mistake. Refactored.

**Pull request merged**